src/base/components/analog_pins.rs

```diff
@@ -0,0 +1,52 @@
+use arduino_nano33iot::hal;
+use hal::adc::Adc;
+use hal::prelude::*;
+use hal::gpio::v2::pin::*;
+use hal::hal::digital::v2::{OutputPin, ToggleableOutputPin, InputPin};
+use super::error::PinError;
+
+pub struct AnalogPin<C> {
+    pub pin: C,
+}
+
+pub trait AnalogInput {
+    fn read(&mut self, adc: &mut Adc<atsamd21g::ADC>) -> Result<u16, PinError>;
+}
+
+macro_rules! is_analog_input {
+    ($d_id:ident, $port:ident) => {
+        impl<'a, I, M> AnalogPin<Pin<I, M>>
+        where I: PinId, M: PinMode {
+            pub fn $d_id(pin: Pin<I, M>) -> AnalogPin<Pin<I, AlternateB>> {
+                let alternate_pin: Pin<I, AlternateB> = pin.into_alternate();
+                AnalogPin{pin: alternate_pin}
+            }
+        }
+
+        impl AnalogInput for AnalogPin<Pin<$port, AlternateB>> {
+            fn read(&mut self, adc: &mut Adc<atsamd21g::ADC>) -> Result<u16, PinError> {
+                match adc.read(&mut self.pin) {
+                    Ok(val) => Ok(val),
+                    Err(_) => Err(PinError::ReadError)
+                }
+            }
+        }
+    };
+}
+
+macro_rules! get_input {
+    ($stone:expr, $port:ident) => {
+        base::components::analog_pins::AnalogPin::$port($stone.pins.$port)
+    };
+}
```

Comment:
I'm not sure I fully understand what's going on here yet. `$port` is used both as a function on `AnalogPin` as well as a field from `$stone` within it? The whole `$d_id` thing above seems overly complex as well.

Can you help me understand what's going on here a bit more so I can try to think of a simpler way to implement this?

Reply:
`$d_id` is used as a function, almost entirely so that if you use a port that I have not whitelisted (below it, via `is_analog_input!` it will throw a compilation error:
https://github.com/nixpulvis/stonecast/pull/6/files/4fe75bb54f8067eba5f9462c75acdbca29b80eae#diff-5c5ba425cbe13407bc70c50a0aad10ebcd149260dec88dd981a21763f6374070R44-R52

So it's essentially there to make sure that if you give it the name of a **digital pin** when you want an **analog input**, it will definitely fail to compile.

Anything more than that would be because each pin, as given to us by the HAL library, has its own Rust Type. So I spent a lot of effort making sure that the types are properly propagated. You might get some type errors if you removed this for a different implementation that uses a single function, though I think I've cleaned a lot of that up in the past couple hours via `impl<X> ... where X: Trait` syntax...